src/Voron/Impl/EncryptionBuffersPool.cs

```diff
@@ -1,125 +1,295 @@
         {
             size = Bits.PowerOf2(size);
 
-            if (size > Constants.Size.Megabyte * 16)
+            if (size > _maxBufferSizeToKeepInBytes)
             {
                 // We don't want to pool large buffers
                 return PlatformSpecific.NativeMemory.Allocate4KbAlignedMemory(size, out thread);
             }
 
             var index = Bits.MostSignificantBit(size);
-
-            if (_items[index].TryPop(out var allocation))
+            while (_items[index].TryPop(out var allocation))
             {
-                thread = allocation.AllocatingThread;
-                return (byte*)allocation.Ptr;
+                if (allocation.InUse.Raise() == false)
+                    continue;
+
+                thread = NativeMemory.ThreadAllocations.Value;
+                thread.Allocations += size;
+                return allocation.Ptr;
             }
 
             return PlatformSpecific.NativeMemory.Allocate4KbAlignedMemory(size, out thread);
         }
 
-        public void Return(byte* ptr, long size, NativeMemory.ThreadStats allocatingThread)
+        public void Return(byte* ptr, long size, NativeMemory.ThreadStats allocatingThread, long generation)
         {
             if (ptr == null)
                 return;
 
             size = Bits.PowerOf2(size);
             Sodium.sodium_memzero(ptr, (UIntPtr)size);
 
-            if (size > Constants.Size.Megabyte * 16 || LowMemoryNotification.Instance.LowMemoryState)
+            if (size > _maxBufferSizeToKeepInBytes || (_isLowMemory.IsRaised() && generation < Generation))
             {
-                // We don't want to pool large buffers / clear them up on low memory
+                // - don't want to pool large buffers
+                // - release all the buffers that were created before we got the low memory event
                 PlatformSpecific.NativeMemory.Free4KbAlignedMemory(ptr, size, allocatingThread);
                 return;
             }
 
+            // updating the thread allocations since we released the memory back to the pool
+            NativeMemory.UpdateMemoryStatsForThread(allocatingThread, size);
+
             var index = Bits.MostSignificantBit(size);
-            _items[index].Push(new NativeAllocation()
+            _items[index].Push(new NativeAllocation
             {
-                Ptr = (IntPtr)ptr,
-                AllocatingThread = allocatingThread,
-                Size = size
+                Ptr = ptr,
+                Size = size,
+                InPoolSince = DateTime.UtcNow
             });
         }
 
-        public void ReleaseUnmanagedResources()
+        public void LowMemory(LowMemorySeverity lowMemorySeverity)
         {
+            if (_isLowMemory.Raise())
+            {
+                Interlocked.Increment(ref _generation);
+            }
+
+            if (_isExtremelyLowMemory.Raise() == false || lowMemorySeverity != LowMemorySeverity.ExtremelyLow)
```

Comment:
You are always raising the extreme low memory

Reply:
done